aesd-char-driver/aesd-circular-buffer.c

```diff
@@ -0,0 +1,191 @@
+        // This is not the normal circular buffer read implementation because we do not want the out_offs value to change
+
+        // Normal condition when the write pointer is ahead of the read pointer
+        if(((buffer->out_offs) + count) < AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED)
+        {
+            // When the current reading node and the in_offs are equal then we have reached the end of the buffer
+            if(((buffer->out_offs) + count) == (buffer->in_offs))
+                break;
+
+            // Add the size of the current node and loop back
+            total_buffer_size += buffer->entry[(buffer->out_offs) + count].size;
+        }
+        // Roll over condition when the write pointer is behind the read pointer
+        else
+        {
+            // When the current reading node and the in_offs are equal then we have reached the end of the buffer
+            if(((buffer->out_offs) + (count-AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED)) == (buffer->in_offs))
+                break;
+
+            // Add the size of the current node and loop back
+            total_buffer_size += buffer->entry[(buffer->out_offs) + (count-AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED)].size;
+        }
+
+        count++;
+    }
+
+    return total_buffer_size;
+}
+
+/**
+ * @param buffer the buffer to search for corresponding offset.  Any necessary locking must be performed by caller.
+ * @param char_offset the position to search for in the buffer list, describing the zero referenced
+ *      character index if all buffer strings were concatenated end to end
+ * @param entry_offset_byte_rtn is a pointer specifying a location to store the byte of the returned aesd_buffer_entry
+ *      buffptr member corresponding to char_offset.  This value is only set when a matching char_offset is found
+ *      in aesd_buffer.
+ * @return the struct aesd_buffer_entry structure representing the position described by char_offset, or
+ * NULL if this position is not available in the buffer (not enough data is written).
+ */
+struct aesd_buffer_entry *aesd_circular_buffer_find_entry_offset_for_fpos(struct aesd_circular_buffer *buffer,
+            size_t char_offset, size_t *entry_offset_byte_rtn )
+{
+    /**
+    * TODO: implement per description
+    */
+
+    uint8_t count = 0;
+    uint32_t total_buffer_size = 0;
+
+    // Obtaining the total size of the elements in the circular buffer
+    total_buffer_size = size_of_aesd_circular_buffer(buffer);
+
+    // If the require character offset char_offset is greater than the total size in the buffer we will return NULL
+    if(char_offset >= total_buffer_size)
+        return NULL;
+    
+    count = 0;
+    // We will loop through and reduce the character offset until we are able to get the offset value in the size
+    // (i.e. no more elements to read)
+    while(count <= AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED)
```

Comment:
@nihalthirunakarasu this works but you could simplify by handling a request for an offset > total size of the buffer and consolidate to one loop.